File: util/role.go
package util

import (
	"github.com/DavidArthurCole/EggLedger/ledgerdata"
)

func RoleFromEB(earningsBonus float64) (string, string, string, float64, int) {

	earningsBonusCopy := earningsBonus
	ooms := 0
	for earningsBonusCopy >= 1e3 && ooms < 17 {
		earningsBonusCopy /= 1e3
		ooms++
	}
	var precision int
	switch {
	case earningsBonusCopy < 10.0:
		precision = 2
	case earningsBonusCopy < 100.0:
		precision = 1
	default:
		precision = 0
	}

	roles := ledgerdata.Config.FarmerRoles
	for _, role := range roles {
		if ((ooms * 3) - precision) == role.OOM {
			//Return role and addendum
			return role.Color, role.Name, Addendum(ooms), earningsBonusCopy, precision
		}
	}
	role := roles[len(roles)-1]
	return role.Color, role.Name, Addendum(roles[len(roles)-1].OOM), earningsBonusCopy, precision
}

File: src/EggLedger.Domain/Util/Role.cs
using EggLedger.Domain.LedgerData;

namespace EggLedger.Domain.Util;

public static class Role {
    public static (string Color, string Name, string Addendum, double Value, int Precision) RoleFromEB(double earningsBonus) {
        var earningsBonusCopy = earningsBonus;
        var ooms = 0;
        while (earningsBonusCopy >= 1e3 && ooms < 17) {
            earningsBonusCopy /= 1e3;
            ooms++;
        }
        int precision;
        if (earningsBonusCopy < 10.0) {
            precision = 2;
        } else if (earningsBonusCopy < 100.0) {
            precision = 1;
        } else {
            precision = 0;
        }

        var roles = LedgerData.LedgerData.Config.FarmerRoles;
        foreach (var role in roles) {
            if ((ooms * 3) - precision == role.Oom) {
                return (role.Color, role.Name, Format.Addendum(ooms), earningsBonusCopy, precision);
            }
        }
        var last = roles[^1];
        return (last.Color, last.Name, Format.Addendum(last.Oom), earningsBonusCopy, precision);
    }
}
